client/web/wasm/src/document.rs

```diff
@@ -30,12 +30,31 @@ pub fn set_tool_options(tool: String, options: &JsValue) -> Result<(), JsValue>
 	match options.into_serde::<ToolOptions>() {
 		Ok(options) => EDITOR_STATE.with(|editor| match translate_tool(&tool) {
 			Some(tool) => editor.borrow_mut().handle_message(ToolMessage::SetToolOptions(tool, options)).map_err(convert_error),
-			None => Err(Error::new(&format!("Couldn't select {} because it was not recognized as a valid tool", tool)).into()),
+			None => Err(Error::new(&format!("Couldn't set options for {} because it was not recognized as a valid tool", tool)).into()),
 		}),
-		Err(err) => Err(Error::new(&format!("Invalud JSON for ToolOptions: {}", err)).into()),
+		Err(err) => Err(Error::new(&format!("Invalid JSON for ToolOptions: {}", err)).into()),
 	}
 }
 
+/// Send a message to a given tool
+#[wasm_bindgen]
+pub fn send_tool_message(tool: String, message: &JsValue) -> Result<(), JsValue> {
+	let tool_message = match translate_tool(&tool) {
+		Some(tool) => match tool {
+			ToolType::Select => match message.into_serde::<tools::select::SelectMessage>() {
+				Ok(select_message) => Ok(ToolMessage::Select(select_message)),
+				Err(err) => Err(Error::new(&format!("Invalid message for {}: {}", tool, err)).into()),
+			},
+			_ => Err(Error::new(&format!("Tool message sending not implemented for {}", tool)).into()),
+		},
+		None => Err(Error::new(&format!("Couldn't send message for {} because it was not recognized as a valid tool", tool)).into()),
+	};
+	EDITOR_STATE.with(|editor| match tool_message {
+		Ok(tool_message) => editor.borrow_mut().handle_message(tool_message).map_err(convert_error),
+		Err(err) => Err(err),
+	})
+}
```

Comment:
What are the cons of just adding a  `dispatch_message(message: &JsValue)` ? (or `dispatch_tool_message`)

Reply:
I chose this approach because it allows the frontend to simply send a message to the active tool, without having to wrap it in `ToolMessage::Select { SelectMessage { ... } }`. For the specific use case of messaging, I think it makes sense to allows the frontend to just specify the name of the message, and the backend to handle wrapping and error handling as granularly as necessary. I feel that most cases of TS -> Rust messaging can and should be specific, otherwise constructing messages on the backend.

I'm not sure what you mean by `dispatch_tool_message`... do you mean I could rename this function, or change it somehow?

Reply:
> I'm not sure what you mean by dispatch_tool_message... do you mean I could rename this function, or change it somehow?
removing one layer of wrapping basically, was just worried about the scalability of that approach but after all thee wasm-wrapper was never intended to be pretty so that should be fine.
Feel free to merge then